### PR TITLE
Add the same compile flags as with ros2_controllers and fix errors

### PR DESCRIPTION
**Pull request by christophfroehlich**
Similar to https://github.com/ros-controls/ros2_controllers/pull/961

@firesurfer as we don't have c++20 on humble, we get an error with the designated initializers. 
And I'm not sure about the line 
```c++
std::optional<DefaultConstructable> rt_data_access = box;
```
which gave the error
```
/workspaces/ros2_rolling_ws/src/realtime_tools/test/realtime_box_best_effort_tests.cpp:133:56: error: choosing ‘constexpr std::optional<_Tp>::optional(_Up&&) [with _Up = realtime_tools::RealtimeBoxBestEffort<DefaultConstructable>&; typename std::enable_if<__and_v<std::__not_<std::is_same<std::optional<_Tp>, typename std::remove_cv<typename std::remove_reference<_Tuple>::type>::type> >, std::__not_<std::is_same<std::in_place_t, typename std::remove_cv<typename std::remove_reference<_Tuple>::type>::type> >, std::is_constructible<_T1, _U1>, std::is_convertible<_U1, _T1> >, bool>::type <anonymous> = true; _Tp = DefaultConstructable]’ over ‘realtime_tools::RealtimeBoxBestEffort<T, mutex_type>::operator std::optional<_Tp>() const [with U = DefaultConstructable; <template-parameter-2-2> = void; T = DefaultConstructable; mutex_type = std::mutex]’ [-Werror=conversion]
  133 |   std::optional<DefaultConstructable> rt_data_access = box;
      |                                                        ^~~
/workspaces/ros2_rolling_ws/src/realtime_tools/test/realtime_box_best_effort_tests.cpp:133:56: error:   for conversion from ‘realtime_tools::RealtimeBoxBestEffort<DefaultConstructable>’ to ‘std::optional<DefaultConstructable>’ [-Werror=conversion]
/workspaces/ros2_rolling_ws/src/realtime_tools/test/realtime_box_best_effort_tests.cpp:133:56: note:   because conversion sequence for the argument is better
cc1plus: some warnings being treated as errors
```
Could be fixed with 
```c++
std::optional<DefaultConstructable> rt_data_access = box.tryGet();
```
or
```c++
std::optional<DefaultConstructable> rt_data_access = static_cast<std::optional<DefaultConstructable>>(box);
```
or maybe with adapting the `std::optional<U>> tryGet() `operator overload?

### CODE REVIEW

**Review by saikishor — APPROVED** [review]
LGTM

File: test/realtime_box_best_effort_tests.cpp [hunk 1]
```diff
@@ -86,7 +86,9 @@ TEST(RealtimeBoxBestEffort, non_default_constructable)
 }
 TEST(RealtimeBoxBestEffort, standard_get)
 {
-  RealtimeBoxBestEffort<DefaultConstructable> box(DefaultConstructable{.a = 1000});
+  DefaultConstructable data_construct;
```
Comment (firesurfer) [inline test/realtime_box_best_effort_tests.cpp:89]: This should also work:

```
RealtimeBoxBestEffort<DefaultConstructable> box(DefaultConstructable{1000});
```

File: test/realtime_box_best_effort_tests.cpp [hunk 2]
```diff
@@ -122,15 +124,18 @@ TEST(RealtimeBoxBestEffort, assignment_operator)
...
+  RealtimeBoxBestEffort box(data_construct);
 
   // Use non RT access
   DefaultConstructable data = box;
 
   EXPECT_EQ(data.a, 100);
 
   // Use RT access -> returns std::nullopt if the mutex could not be locked
-  std::optional<DefaultConstructable> rt_data_access = box;
+  std::optional<DefaultConstructable> rt_data_access = box.tryGet();
```
Comment (firesurfer) [inline test/realtime_box_best_effort_tests.cpp:138]: I am a bit puzzled why this doesn't work. I don't remember seeing a warning before on that line. (Which is now escalated to an error)

File: test/realtime_box_best_effort_tests.cpp [hunk 3]
```diff
@@ -86,7 +86,9 @@ TEST(RealtimeBoxBestEffort, non_default_constructable)
 }
 TEST(RealtimeBoxBestEffort, standard_get)
 {
-  RealtimeBoxBestEffort<DefaultConstructable> box(DefaultConstructable{.a = 1000});
+  DefaultConstructable data_construct;
+  data_construct.a = 1000;
+  RealtimeBoxBestEffort<DefaultConstructable> box(data_construct);
```
Comment (bmagyar) [inline test/realtime_box_best_effort_tests.cpp:91]:
```suggestion
  RealtimeBoxBestEffort<DefaultConstructable> box(DefaultConstructable{1000});
```
let's try

File: test/realtime_box_best_effort_tests.cpp (hunk 2 above)
Comment (christophfroehlich) [reply test/realtime_box_best_effort_tests.cpp:136]: Maybe is this warning not included in `-Wall -Wextra`?

Comment (saikishor) [reply test/realtime_box_best_effort_tests.cpp:136]: I believe what @christophfroehlich did here is the right approach. The problem is that the box is of type `RealtimeBoxBestEffort<T>` and we are initializing std::optional<T> with this, so when using flag `-Werror=conversion` this fails

**Review by firesurfer — APPROVED** [review]

**Review by saikishor — APPROVED** [review]

**Pull request merged**